### ISSUE TITLE
[helm] Upgrade from 2.3.4 to 2.4.0 is not stable

### ISSUE THREAD

**Issue by anjmao** [nats-io/k8s#312]
Just tried to upgrade from 2.3.4 to 2.4.0 version via helm chart.

```yaml
  nats:
    image: nats:2.4.0-alpine
    resources:
      requests:
        cpu: 1
        memory: 2Gi
      limits:
        cpu: 2
        memory: 2Gi

    limits:
      maxConnections:
      maxSubscriptions:
      maxControlLine:
      maxPayload:
      writeDeadline:
      maxPending:
      maxPings:
      lameDuckDuration:

    terminationGracePeriodSeconds: 60

    logging:
      debug:
      trace:
      logtime:
      connectErrorReports:
      reconnectErrorReports:

    jetstream:
      enabled: true

      memStorage:
        enabled: true

      fileStorage:
        enabled: true
        storageDirectory: /data
        storageClassName: standard
        size: 30Gi
        accessModes:
          - ReadWriteOnce

  cluster:
    enabled: true
    replicas: 3
    noAdvertise: false

  auth:
    enabled: true
    systemAccount:
    resolver:
      type: memory
      configMap:
        name: nats-accounts
        key: resolver.conf

  natsbox:
    enabled: false

  exporter:
    enabled: true   
```

Upgrade completely failed and streams stopped responding. Issues:

```
[ERR] RAFT [vURD0sM3 - S-R3F-pU5w3rId] Error sending snapshot to follower [yLCaJyhQ]: raft: no snapshot available
```

```
nats: no response from stream
```

No leader on stream. Also saw duplicate streams.
```
nats-js-0!, nats-js-1!, nats-js-2!
``` 

Also I tried to purge some problematic streams or change leader eg: `nats stream cluster step-down agent` but only received errors or timeouts.

**Comment by anjmao** [comment]
Hi @wallyqs 

We have:
* 6 streams.
* 25 msg/s publish rate.
* 14 consumers.

I compared helm chart diff as we there using 0.8.2 chart version and didn't saw any changes to directory paths. However I think found something else:

Currently we are using `github.com/nats-io/nats.go v1.11.0` client version. It looks that 2.4.0 server is not compatible. Here is example code which I ran inside k8s cluster.

```go
func main() {
	log := logrus.New()

	connOpts := []nats.Option{
		nats.MaxReconnects(-1),
		nats.ReconnectWait(10 * time.Second),
		nats.DisconnectErrHandler(func(_ *nats.Conn, err error) {
			if err != nil {
				log.Errorf("nats client disconnected: %v", err)
			} else {
				log.Infof("nats client disconnected")
			}
		}),
		nats.ReconnectHandler(func(_ *nats.Conn) {
			log.Info("nats client reconnected")
		}),
		nats.ClosedHandler(func(_ *nats.Conn) {
			log.Info("nats client closed")
		}),
		nats.ErrorHandler(func(conn *nats.Conn, subscription *nats.Subscription, err error) {
			log.Errorf("nats client error: %v", err)
		}),
		nats.UserCredentials("/etc/nats-creds/kubecast.creds"),
	}

	nc, err := nats.Connect(
		"nats://nats-js-0.nats-js:4222, nats://nats-js-1.nats-js:4222, nats://nats-js-2.nats-js:4222",
		connOpts...,
	)
	if err != nil {
		log.Fatal(err)
	}

	js, err := nc.JetStream()
	if err != nil {
		log.Fatal(err)
	}

	_, _ = js.AddStream(&nats.StreamConfig{
		Name:     "foo",
		Subjects: []string{"foo.*"},
		Storage:  nats.FileStorage,
		Replicas: 3,
		MaxAge:   24 * time.Hour,
	})

	for i := 0; i < 10; i++ {
		queueName := fmt.Sprintf("q%d", i)
		_, err := js.QueueSubscribe("foo.created", queueName, func(msg *nats.Msg) {
			log.Infof("received msg to %s", queueName)
			if err := msg.Ack(); err != nil {
				return
			}
		},
			nats.Durable(fmt.Sprintf("foo_%s", queueName)),
			nats.AckWait(1*time.Minute),
			nats.MaxDeliver(10),
			nats.DeliverLast(),
			nats.ManualAck())

		if err != nil {
			log.Fatal(err)
		}
	}

	for {
		msg := []byte(fmt.Sprintf("hello %d", time.Now().UnixNano()))
		_, err := js.Publish("foo.created", msg)
		if err != nil {
			log.Errorf("publish failed: %v", err)
		}
		log.Infof("published: %s", string(msg))
		time.Sleep(1 * time.Second)
	}
}
```

* After the upgrade to 2.4.0 all consumers stopped receiving messages. Pod restart didn't helped.
* Only after upgrading to newest go client v1.12.0 I saw messages being processed again.

The important part is that `v1.12.0` client is not compatible with `v2.3.4` server as it requires DeliveryGroup. And `2.4.0` server doesn't work with `v1.11.0` client correctly. So there is no upgrade path without downtime.

Please update nats-server release notes and mention this.

**Comment by anjmao** [comment]
Also It looks that users will lose data:

* After upgraded to `2.4.0` server and later `v1.12.0` client nats client connection always crashed with `nats: JetStream system temporarily unavailable` error.

* On server side 
 `JetStream cluster consumer 'AAQE7F2XQY6PODIDHXM6F5MCHTTTYF4N4Z6O6SFGOAA3HCVUHC7J4L4R > foo > foo_q4' has NO quorum, stalled`

Only after recreating the stream I got it working.

**Comment by ColinSullivan1** [comment]
@anjmao , Thank you for the update, much appreciated.  We're sorry about your troubles upgrading.

We'll provide updates or guidance for upgrading around this particular version and will work to ensure compatibility or provide better guidance around potential upgrade issues moving forward.

**Comment by ColinSullivan1** [comment]
FYI, We've added some detail in the [NATS server 4.2.0 release notes](https://github.com/nats-io/nats-server/releases/tag/v2.4.0).  We will be better about this moving forward.

**Comment by wallyqs** [comment]
Closing since this should be much better now with the recovery fixes v2.6.1 out, and also increased the `terminationGracePeriodSeconds` to be at least 2min as the lame duck mode duration.
